fms/utils/spyre/paged.py

```diff
@@ -0,0 +1,524 @@
+        values_lst: list[torch.Tensor] = []
+        for j in range(start_pos, seq_len):
+            block_number = int(block_table[j // block_size])
+            block_offset = j % block_size
+
+            k = key_cache[block_number, block_offset, :, :]
+            k = k.reshape(num_kv_heads, head_size)
+            keys_lst.append(k)
+
+            v = value_cache[block_number, block_offset, :, :]
+            values_lst.append(v)
+        keys = torch.stack(keys_lst, dim=0)
+        values = torch.stack(values_lst, dim=0)
+        if num_kv_heads > 1:
+            # Handle MQA and GQA
+            keys = torch.repeat_interleave(keys, num_query_heads // num_kv_heads, dim=1)
+            values = torch.repeat_interleave(
+                values, num_query_heads // num_kv_heads, dim=1
+            )
+
+        out = ref_masked_attention(q, keys, values, scale)
+        out = out.view(num_query_heads, head_size)
+        output[i].copy_(out, non_blocking=True)
+    return output
+
+
+@paged_attn_compute.register_fake
+def paged_attn_compute_meta(
+    query: torch.Tensor,
+    key_cache: torch.Tensor,
+    value_cache: torch.Tensor,
+    scale: float,
+    current_tkv_mask: torch.Tensor,
+    left_padded_prompt_mask: torch.Tensor,
+    block_table: torch.Tensor,
+) -> torch.Tensor:
+    return torch.zeros_like(query)
+
+
+class SpyrePagedAttentionKwargs(AttentionKwargs):
+    current_tkv_mask: Optional[torch.Tensor]
+    left_padded_prompt_mask: Optional[torch.Tensor]
+    block_table: Optional[torch.Tensor]
+    slot_mapping: torch.Tensor
+    mask: Optional[torch.Tensor]  # prefill mask
+
+
+def __spyre_paged_store_op(
+    keys: torch.Tensor,
+    values: torch.Tensor,
+    key_cache: Optional[torch.Tensor],
+    value_cache: Optional[torch.Tensor],
+    **attn_kwargs,
+):
+    result_key_cache, result_value_cache = torch.ops.spyre.paged_attn_store(
+        keys, values, key_cache, value_cache, attn_kwargs["slot_mapping"]
+    )
+
+    # for prefill, we want to return the original keys/values
+    if attn_kwargs.get("block_table", None) is None:
+        return keys, values, result_key_cache, result_value_cache
+    else:
+        return (
+            result_key_cache,
+            result_value_cache,
+            result_key_cache,
+            result_value_cache,
+        )
+
+
+def __spyre_paged_compute_op(
+    query: torch.Tensor,
+    key_cache: torch.Tensor,
+    value_cache: torch.Tensor,
+    nheads: int,
+    kvheads: int,
+    p_dropout: float,
+    scale_factor: Optional[float],
+    **attn_kwargs,
+):
+    if scale_factor is None:
+        scale_factor = 1 / math.sqrt(query.shape[-1])
+    return torch.ops.spyre.paged_attn_compute(
+        query,
+        key_cache,
+        value_cache,
+        scale_factor,
+        attn_kwargs["current_tkv_mask"],
+        attn_kwargs["left_padded_prompt_mask"],
+        attn_kwargs["block_table"],
+    )
+
+
+def __spyre_paged_validate_attn_kwargs_op(
```

Comment:
add documentation for this new function as well in the general interface for attention